qprofiler2/src/org/qcmg/qprofiler2/summarise/PairSummary.java

```diff
@@ -0,0 +1,220 @@
+   			result = Math.min( read_end, mate_end ) - Math.max(record.getAlignmentStart(), record.getMateAlignmentStart() ) + 1;   
+   		}
+   		
+   		//|--->    or     |-----> (read end - mate start>0) or       |----->
+   		//|----->              |----->                      |---> (read end - mate start < 0)  		
+		//if tLen >= 0, then mate_start > read_start; so min_end will be read end if we assue pair with same length. 
+   		else if( !record.getReadNegativeStrandFlag()) {  		
+   			result = record.getAlignmentEnd() - record.getMateAlignmentStart() + 1; 			
+   		}else {  		
+	  		//<---|    or     <-----| (read end - mate start>0) or       <-----|
+	   		//<-----|             <-----|                      <---| (read end - mate start < 0)
+	   		result = record.getAlignmentEnd() - Math.max( record.getAlignmentStart(), record.getMateAlignmentStart()) + 1;			
+   		}
+   		
+   		return result; 	
+    }
+	public static Pair getPairType( SAMRecord  record ) {
+	 
+		if(record.getReadUnmappedFlag() ||  record.getMateUnmappedFlag() || ! record.getReferenceName().equals(record.getMateReferenceName()) )
+			return Pair.Others;
+		
+		// pair with different orientation 
+		if( record.getReadNegativeStrandFlag() != record.getMateNegativeStrandFlag()  ) {	
+			// |----->  <-------|(read is reverse)
+			if( record.getReadNegativeStrandFlag() && record.getAlignmentStart() >= record.getMateAlignmentStart())
+				return Pair.Inward;
+				
+			// (read is forward)|----->  <-------| 
+			if( !record.getReadNegativeStrandFlag() && record.getAlignmentStart() <= record.getMateAlignmentStart())
+				return Pair.Inward;
+	 					
+			return Pair.Outward;
+		}
+				
+		//non-canonical: pair with same orientation  		
+		//first of pair start
+		int s1 = ( record.getFirstOfPairFlag())? record.getAlignmentStart() : record.getMateAlignmentStart();		
+		//second of pair start
+		int s2 = ( !record.getFirstOfPairFlag())? record.getAlignmentStart() : record.getMateAlignmentStart();
+						    	 		 
+		//F3(s1)|---->  F5(s2)|------>
+		if( !record.getReadNegativeStrandFlag() &&  s1 <= s2)  return Pair.F3F5;
+		//F5(s2)<----|  F3(s1)<------|
+		if( record.getReadNegativeStrandFlag() &&  s1 >= s2)  return Pair.F3F5;		
+		//F5(s2)|---->  F3(s1)|------>
+		if( !record.getReadNegativeStrandFlag() &&  s1 > s2)  return Pair.F5F3;
+		//F3(s1)<----|  F5(s2)<----|
+		if( record.getReadNegativeStrandFlag() &&  s1 < s2)  return Pair.F5F3;
+				
+		return Pair.Others;
+	}	
+	
+	public final Pair type;
+	public final Boolean isProperPair; 
+	PairSummary( Pair pair, boolean isProper){this.type = pair; this.isProperPair = isProper;}
+	
+	@Override
+	public boolean equals(Object obj) {
+		if (obj == null) return false;		
+		if (!(obj instanceof PairSummary)) return false;		
+		if(obj.hashCode() == this.hashCode() ) return true;
+		
+		return false;
+	}
+	
+	@Override
+	public int hashCode() {	
+	    return isProperPair? this.type.id * 100 : this.type.id * 1;	    
```

Comment:
is this hashCode function ok? Have you tested how it behaves?

Reply:
I should consider the case when the type variable is null.  But anyway, I decide not to override equals and hashCode method. The original plan is to create unique PairSummary based on pair orientation and ProperPair flag. However, two objects may have the same Pair and flag but still possible have different counts. They should be considered non-equal.